### PR TITLE
Check allowed page types

### LINKED ISSUE #1521: Mehrere Fehlerseiten 404 unter einem Startpunkt
**Affected version(s)**

Festgestellt in Contao 4.9

**Description**

Unter einem Startpunkt können mehrere Fehlerseiten 404 (Seitentyp "404 Seite nicht gefunden") angelegt werden. Es wird dann offenbar stets die erste Seite angezeigt.

So wie unter einem Startpunkt nicht mehrere Seiten mit demselben Alias angelegt werden können, sollte es jedoch auch nicht möglich sein, mehrere Fehlerseiten 404 unter einem Startpunkt anzulegen.

### PR DESCRIPTION
**Pull request by aschempp**
| Q                | A
| -----------------| ---
| Fixed issues     | Fixes #1521

With this PR Contao only allows acceptable page type options in the respective place:

1. The root node can only contain `root` (already present)
2. `root` is not allowed with a parent page
3. `error_401`, `error_403` and `error_404` are only allowed as direct descendents of type `root`
4. `error_401`, `error_403` and `error_404` can only exist once per root page

This PR is against Contao 4.10, because it added the feature to limit the page type options. I don't think we should invest into fixing this in Contao 4.9, also because it might introduce a slightly different behavior for people.

I'm not 100% sure if the error pages can only exist once per root page? What about permissions or unpublished pages? Also, this is not 100%, it does not prevent copying the page type etc., but I don't think it should.

### CODE REVIEW

File: core-bundle/src/EventListener/FilterPageTypeListener.php [hunk 1]
```diff
@@ -27,11 +39,33 @@ public function __invoke(FilterPageTypeEvent $event): void
...
+        // The first level can only have root pages (see #6360)
+        if (!$dc->activeRecord->pid) {
             $event->setOptions(['root']);
+
+            return;
+        }
+
+        $event->removeOption('root');
+
+        $parentType = $this->connection->fetchColumn('SELECT type FROM tl_page WHERE id=?', [$dc->activeRecord->pid]);
```
Comment (leofeyer) [inline core-bundle/src/EventListener/FilterPageTypeListener.php:51]: The method is deprecated in favor of the `fetchOne()` method. Did you check if we can use the new method?

File: core-bundle/src/EventListener/FilterPageTypeListener.php [hunk 2]
```diff
@@ -27,11 +39,33 @@ public function __invoke(FilterPageTypeEvent $event): void
...
+            $event->removeOption('error_401');
+            $event->removeOption('error_403');
+            $event->removeOption('error_404');
+
+            return;
+        }
+
+        $siblingTypes = $this->connection
+            ->executeQuery('SELECT DISTINCT(type) FROM tl_page WHERE pid=?', [$dc->activeRecord->pid])
+            ->fetchAll(FetchMode::COLUMN)
```
Comment (leofeyer) [inline core-bundle/src/EventListener/FilterPageTypeListener.php:64]: The method is deprecated in favor of the `fetchFirstColumn()` method. Did you check if we can use the new method?

**Comment by leofeyer** [comment]
> I'm not 100% sure if the error pages can only exist once per root page? What about permissions or unpublished pages?

Contao does not check user permissions when looking for an error page, therefore one error page per root should be fine.

> Also, this is not 100%, it does not prevent copying the page type etc., but I don't think it should.

I think it should. 🙈 @contao/developers WDYT?

**Comment by aschempp** [comment]
I don't think it should for multiple reasons. I might want to make a copy to prepare a new error page, and unpublish the existing one after that. Hiding the options is a _hint_ on what you _should_ do, but same as you can copy a root page into a root page, this should be allowed as well.

**Comment by leofeyer** [comment]
Hm, I kind of agree. 😄

**Review by leofeyer — APPROVED** [review]

**Pull request merged**